src/autobattle.cpp

```diff
@@ -0,0 +1,339 @@
+#include "autobattle.h"
+#include "game_actor.h"
+#include "game_enemy.h"
+#include "game_enemyparty.h"
+#include "game_party.h"
+#include "game_battlealgorithm.h"
+#include "game_battle.h"
+#include "algo.h"
+#include "player.h"
+#include "output.h"
+#include <lcf/reader_util.h>
+#include <lcf/data.h>
+
+namespace AutoBattle {
+
+#ifdef EP_DEBUG_AUTOBATTLE
+template <typename... Args>
+static void DebugLog(const char* fmt, Args&&... args) {
+	Output::Debug(fmt, std::forward<Args>(args)...);
+}
+#else
+template <typename... Args>
+static void DebugLog(const char*, Args&&...) {}
+#endif
+
+constexpr decltype(RpgRtCompat::name) RpgRtCompat::name;
+constexpr decltype(AttackOnly::name) AttackOnly::name;
+constexpr decltype(RpgRtImproved::name) RpgRtImproved::name;
+
+std::unique_ptr<AlgorithmBase> CreateAlgorithm(StringView name) {
+	if (name == RpgRtImproved::name) {
+		return std::make_unique<RpgRtImproved>();
+	}
+	if (name == AttackOnly::name) {
+		return std::make_unique<AttackOnly>();
+	}
+	return std::make_unique<RpgRtCompat>();
+}
+
+void AlgorithmBase::SetAutoBattleAction(Game_Actor& source) {
+	vSetAutoBattleAction(source);
+	if (source.GetBattleAlgorithm() == nullptr) {
+		source.SetBattleAlgorithm(std::make_shared<Game_BattleAlgorithm::NoMove>(&source));
+	}
+}
+
+void RpgRtCompat::vSetAutoBattleAction(Game_Actor& source) {
+	SelectAutoBattleActionRpgRtCompat(source, Game_Battle::GetBattleCondition());
+}
+
+void AttackOnly::vSetAutoBattleAction(Game_Actor& source) {
+	SelectAutoBattleAction(source, Game_Battler::WeaponAll, Game_Battle::GetBattleCondition(), false, false, false, false);
+}
+
+void RpgRtImproved::vSetAutoBattleAction(Game_Actor& source) {
+	SelectAutoBattleAction(source, Game_Battler::WeaponAll, Game_Battle::GetBattleCondition(), true, false, false, false);
+}
+
+double CalcSkillHealAutoBattleTargetRank(const Game_Actor& source, const Game_Battler& target, const lcf::rpg::Skill& skill, bool apply_variance, bool emulate_bugs) {
+	assert(skill.type == lcf::rpg::Skill::Type_normal || skill.type >= lcf::rpg::Skill::Type_subskill);
+	assert(skill.scope == lcf::rpg::Skill::Scope_self || skill.scope == lcf::rpg::Skill::Scope_ally || skill.scope == lcf::rpg::Skill::Scope_party);
+
+	const auto src_max_sp = source.GetMaxSp();
+	const auto tgt_max_hp = target.GetMaxHp();
+	const auto tgt_hp = target.GetHp();
+
+	if (target.GetHp() > 0) {
+		// Can the skill can heal the target?
+		if (!skill.affect_hp) {
+			return 0.0;
+		}
+
+		const auto base_effect = Algo::CalcSkillEffect(source, target, skill, apply_variance);
+		const auto max_effect = std::min(base_effect, tgt_max_hp - tgt_hp);
+
+		auto rank = static_cast<double>(max_effect) / static_cast<double>(tgt_max_hp);
+		if (src_max_sp > 0) {
+			const auto cost = source.CalculateSkillCost(skill.ID);
+			rank -= cost / src_max_sp * 0.125;
```

Comment:
``cost / src_max_sp`` int div in floating point. Intentional?

Reply:
all bugs, thanks!